File: crates/parquet-lab/src/prune.rs
//! Skipping data: deciding, from metadata alone, what a query does not need to read (ch09).
//!
//! A reader is given a condition on one column, such as `order_id = 431` or `amount_cents > 9000`.
//! Before it reads a page it asks three things, each cheaper than the reading it may save:
//!
//! 1. **Row group statistics** ([`crate::stats`]), already in the footer: can any value in this
//!    column chunk's range satisfy the condition?
//! 2. **The Bloom filter** ([`crate::bloom`]), for equality only: was this exact value ever added?
//! 3. **The page index** ([`crate::page_index`]): the same question as the first, page by page,
//!    and then which rows those pages hold, and so which pages of the *other* columns to read.
//!
//! Every answer is either "skip", which must be certain, or "read", which only means "cannot rule
//! it out". A wrong "skip" loses rows; a wrong "read" costs bytes. Each decision records why.

use std::cmp::Ordering;

use crate::bytes::Span;
use crate::metadata::{ColumnChunk, FileMetaData};
use crate::schema::Leaf;
use crate::stats::{bounds, Comparator};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Op {
    Eq,
    NotEq,
    Lt,
    LtEq,
    Gt,
    GtEq,
    IsNull,
    IsNotNull,
}

impl Op {
    pub fn parse(s: &str) -> Result<Op, String> {
        Ok(match s {
            "=" => Op::Eq,
            "!=" => Op::NotEq,
            "<" => Op::Lt,
            "<=" => Op::LtEq,
            ">" => Op::Gt,
            ">=" => Op::GtEq,
            "is null" => Op::IsNull,
            "is not null" => Op::IsNotNull,
            other => return Err(format!("unknown comparison {other:?}")),
        })
    }

    pub fn symbol(&self) -> &'static str {
        match self {
            Op::Eq => "=",
            Op::NotEq => "!=",
            Op::Lt => "<",
            Op::LtEq => "<=",
            Op::Gt => ">",
            Op::GtEq => ">=",
            Op::IsNull => "is null",
            Op::IsNotNull => "is not null",
        }
    }

    /// Whether a value compares this way. Null matches nothing but `is null`.
    pub fn matches(&self, ordering: Option<Ordering>) -> bool {
        use Ordering::*;
        matches!(
            (self, ordering),
            (Op::Eq, Some(Equal))
                | (Op::NotEq, Some(Less | Greater))
                | (Op::Lt, Some(Less))
                | (Op::LtEq, Some(Less | Equal))
                | (Op::Gt, Some(Greater))
                | (Op::GtEq, Some(Greater | Equal))
        )
    }
}

/// A condition on one column: `column op value`.
#[derive(Clone, Debug, PartialEq)]
pub struct Predicate {
    pub column: usize,
    pub op: Op,
    /// The value as PLAIN bytes, as statistics store it. `None` for `is null` and `is not null`.
    pub value: Option<Vec<u8>>,
    pub comparator: Comparator,
}

impl Predicate {
    /// Parse a condition's value for `leaf`: integers, floats and strings.
    pub fn new(
        leaf: &Leaf,
        converted: Option<&str>,
        op: Op,
        text: &str,
    ) -> Result<Predicate, String> {
        let comparator = Comparator::for_leaf(leaf, converted);
        let value = match op {
            Op::IsNull | Op::IsNotNull => None,
            _ => Some(literal(comparator, text.trim())?),
        };
        Ok(Predicate {
            column: leaf.column,
            op,
            value,
            comparator,
        })
    }

    /// Whether one row's value satisfies the condition. `None` is null.
    pub fn row_matches(&self, value: Option<&[u8]>) -> bool {
        match (self.op, value, &self.value) {
            (Op::IsNull, v, _) => v.is_none(),
            (Op::IsNotNull, v, _) => v.is_some(),
            (op, Some(v), Some(x)) => op.matches(self.comparator.compare(v, x)),
            _ => false,
        }
    }
}

/// A literal as PLAIN bytes for a column compared with `c`.
fn literal(c: Comparator, text: &str) -> Result<Vec<u8>, String> {
    let bad = || format!("{text:?} is not a value of this column");
    Ok(match c {
        Comparator::I32 => text
            .parse::<i32>()
            .map_err(|_| bad())?
            .to_le_bytes()
            .to_vec(),
        Comparator::U32 => text
            .parse::<u32>()
            .map_err(|_| bad())?
            .to_le_bytes()
            .to_vec(),
        Comparator::I64 => text
            .parse::<i64>()
            .map_err(|_| bad())?
            .to_le_bytes()
            .to_vec(),
        Comparator::U64 => text
            .parse::<u64>()
            .map_err(|_| bad())?
            .to_le_bytes()
            .to_vec(),
        Comparator::F32 => text
            .parse::<f32>()
            .map_err(|_| bad())?
            .to_le_bytes()
            .to_vec(),
        Comparator::F64 => text
            .parse::<f64>()
            .map_err(|_| bad())?
            .to_le_bytes()
            .to_vec(),
        Comparator::Bytes => text.trim_matches('"').as_bytes().to_vec(),
        _ => {
            return Err(
                "this reader parses values only for integer, float and string columns".into(),
            )
        }
    })
}

/// A decision and the reason for it.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Decision {
    pub skip: bool,
    pub why: String,
}

fn skip(why: impl Into<String>) -> Decision {
    Decision {
        skip: true,
        why: why.into(),
    }
}

fn read(why: impl Into<String>) -> Decision {
    Decision {
        skip: false,
        why: why.into(),
    }
}

/// Decide from a minimum and maximum (when usable), a null count, and a value count.
pub fn against_bounds(
    p: &Predicate,
    bounds: Option<(&[u8], &[u8])>,
    null_count: Option<i64>,
    num_values: i64,
) -> Decision {
    let all_null = null_count == Some(num_values) && num_values > 0;
    match p.op {
        Op::IsNull if null_count == Some(0) => return skip("the null count is zero"),
        Op::IsNull => return read("the null count is not zero, or not recorded"),
        Op::IsNotNull if all_null => return skip("every value is null"),
        Op::IsNotNull => return read("some values are not null"),
        _ if all_null => return skip("every value is null, and null satisfies no comparison"),
        _ => {}
    }
    let (Some(x), Some((min, max))) = (&p.value, bounds) else {
        return read("no usable minimum and maximum");
    };
    let c = p.comparator;
    let (vs_min, vs_max) = match (c.compare(x, min), c.compare(x, max)) {
        (Some(a), Some(b)) => (a, b),
        _ => return read("the value cannot be placed in the column's order"),
    };
    use Ordering::*;
    let ruled_out = match p.op {
        Op::Eq => vs_min == Less || vs_max == Greater,
        Op::NotEq => vs_min == Equal && vs_max == Equal,
        Op::Lt => vs_min != Greater,   // min >= x
        Op::LtEq => vs_min == Less,    // min > x
        Op::Gt => vs_max != Less,      // max <= x
        Op::GtEq => vs_max == Greater, // max < x
        Op::IsNull | Op::IsNotNull => unreachable!(),
    };
    if ruled_out {
        skip(match p.op {
            Op::Eq if vs_min == Less => "the value is below the minimum",
            Op::Eq => "the value is above the maximum",
            Op::NotEq => "every value equals it",
            Op::Lt | Op::LtEq => "the minimum is too large",
            _ => "the maximum is too small",
        })
    } else {
        read("the range may hold a match")
    }
}

/// Which mechanisms a plan may use.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Mechanisms {
    pub statistics: bool,
    pub bloom: bool,
    pub page_index: bool,
}

impl Mechanisms {
    pub const ALL: Mechanisms = Mechanisms {
        statistics: true,
        bloom: true,
        page_index: true,
    };
    pub const NONE: Mechanisms = Mechanisms {
        statistics: false,
        bloom: false,
        page_index: false,
    };
}

/// One page of the condition's column, and whether its rows are kept.
#[derive(Clone, Debug, PartialEq)]
pub struct PagePlan {
    pub span: Span,
    pub rows: (i64, i64),
    pub decision: Decision,
}

/// The bytes of one projected column a plan reads in one row group.
#[derive(Clone, Debug, PartialEq)]
pub struct ColumnRead {
    pub column: usize,
    pub spans: Vec<Span>,
    /// How many of the chunk's data pages those spans include, and how many it has.
    pub pages_read: usize,
    pub pages_total: usize,
}

#[derive(Clone, Debug, PartialEq)]
pub struct RowGroupPlan {
    pub index: usize,
    pub num_rows: i64,
    /// Each mechanism consulted, in order, and what it decided.
    pub steps: Vec<(&'static str, Decision)>,
    pub skipped: bool,
    pub pages: Vec<PagePlan>,
    /// The rows still to read, as `[start, end)` ranges within the row group.
    pub rows: Vec<(i64, i64)>,
    pub reads: Vec<ColumnRead>,
    /// Bytes of Bloom filters and page indexes fetched to decide.
    pub index_bytes: u64,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Plan {
    pub row_groups: Vec<RowGroupPlan>,
}

impl Plan {
    pub fn bytes_read(&self) -> u64 {
        self.row_groups
            .iter()
            .flat_map(|g| &g.reads)
            .flat_map(|r| &r.spans)
            .map(|s| s.len())
            .sum()
    }

    pub fn index_bytes(&self) -> u64 {
        self.row_groups.iter().map(|g| g.index_bytes).sum()
    }

    pub fn rows_read(&self) -> i64 {
        self.row_groups
            .iter()
            .flat_map(|g| &g.rows)
            .map(|(a, b)| b - a)
            .sum()
    }
}

fn overlaps(a: (i64, i64), b: (i64, i64)) -> bool {
    a.0 < b.1 && b.0 < a.1
}

/// Merge touching or overlapping ranges.
fn merge(mut ranges: Vec<(i64, i64)>) -> Vec<(i64, i64)> {
    ranges.sort();
    let mut out: Vec<(i64, i64)> = Vec::new();
    for r in ranges {
        match out.last_mut() {
            Some(last) if r.0 <= last.1 => last.1 = last.1.max(r.1),
            _ => out.push(r),
        }
    }
    out
}

/// The bytes of `chunk` needed for `rows`: its dictionary page, if any, and the data pages
/// whose rows overlap them. Without an offset index, the whole chunk.
fn column_read(
    file: &[u8],
    chunk: &ColumnChunk,
    column: usize,
    rows: &[(i64, i64)],
    num_rows: i64,
    use_index: bool,
) -> (ColumnRead, u64) {
    let whole = chunk.byte_range();
    let oi = if use_index {
        crate::page_index::offset_index(file, chunk).ok().flatten()
    } else {
        None
    };
    let Some(oi) = oi else {
        let pages = crate::page_index::offset_index(file, chunk)
            .ok()
            .flatten()
            .map(|o| o.pages.len())
            .unwrap_or(0);
        return (
            ColumnRead {
                column,
                spans: vec![whole],
                pages_read: pages,
                pages_total: pages,
            },
            0,
        );
    };
    let mut spans = Vec::new();
    // The dictionary page sits before the first data page; any page read needs it.
    let first_data = oi
        .pages
        .first()
        .map(|p| p.offset as u64)
        .unwrap_or(whole.end);
    let ranges = oi.row_ranges(num_rows);
    let wanted: Vec<usize> = (0..oi.pages.len())
        .filter(|&i| rows.iter().any(|&r| overlaps(ranges[i], r)))
        .collect();
    if !wanted.is_empty() && first_data > whole.start {
        spans.push(Span::new(whole.start, first_data));
    }
    spans.extend(wanted.iter().map(|&i| oi.pages[i].span()));
    (
        ColumnRead {
            column,
            spans,
            pages_read: wanted.len(),
            pages_total: oi.pages.len(),
        },
        oi.span.len(),
    )
}

/// Plan a read of `projection` for rows satisfying `p`.
pub fn plan(
    file: &[u8],
    md: &FileMetaData,
    leaf: &Leaf,
    p: &Predicate,
    projection: &[usize],
    use_: Mechanisms,
) -> Result<Plan, String> {
    let type_order = md
        .column_orders
        .as_ref()
        .and_then(|o| o.get(leaf.column))
        .is_some_and(|o| o == "TYPE_ORDER");
    let mut groups = Vec::new();
    for (g, rg) in md.row_groups.iter().enumerate() {
        let chunk = &rg.columns[leaf.column];
        let mut steps = Vec::new();
        let mut index_bytes = 0;
        let mut skipped = false;
        if use_.statistics {
            let d = match &chunk.statistics {
                Some(s) => {
                    let b = bounds(s, p.comparator, type_order).ok();
                    against_bounds(
                        p,
                        b.as_ref().map(|b| (&b.min[..], &b.max[..])),
                        s.null_count,
                        chunk.num_values,
                    )
                }
                None => read("the column chunk has no statistics"),
            };
            skipped |= d.skip;
            steps.push(("statistics", d));
        }
        if use_.bloom && !skipped && p.op == Op::Eq {
            if let (Some(filter), Some(x)) = (crate::bloom::read(file, chunk)?, &p.value) {
                index_bytes += filter.bitset_span.end - filter.header_span.start;
                let probe = filter.probe(x);
                let d = if probe.may_contain {
                    read("every bit the value's hash chose is set")
                } else {
                    skip("a bit the value's hash chose is clear")
                };
                skipped |= d.skip;
                steps.push(("Bloom filter", d));
            }
        }
        let mut pages = Vec::new();
        let mut rows = if skipped {
            vec![]
        } else {
            vec![(0, rg.num_rows)]
        };
        // Page rows equal page values only when the column does not repeat; for a repeated
        // column the reader would need the level histograms, and this one does not use them.
        let use_pages = use_.page_index && !skipped && leaf.max_repetition_level == 0;
        if use_pages {
            let ci = crate::page_index::column_index(file, chunk)?;
            let oi = crate::page_index::offset_index(file, chunk)?;
            if let (Some(ci), Some(oi)) = (ci, oi) {
                index_bytes += ci.span.len();
                let ranges = oi.row_ranges(rg.num_rows);
                for (i, loc) in oi.pages.iter().enumerate() {
                    let page_rows = ranges[i].1 - ranges[i].0;
                    let nulls = ci.null_counts.as_ref().map(|n| n[i]);
                    let b = (!ci.null_pages[i] && type_order)
                        .then(|| (&ci.min_values[i][..], &ci.max_values[i][..]))
                        .filter(|(lo, hi)| {
                            p.comparator.compare(lo, lo).is_some()
                                && p.comparator.compare(hi, hi).is_some()
                        });
                    let nulls = if ci.null_pages[i] {
                        Some(page_rows)
                    } else {
                        nulls
                    };
                    let decision = against_bounds(p, b, nulls, page_rows);
                    pages.push(PagePlan {
                        span: loc.span(),
                        rows: ranges[i],
                        decision,
                    });
                }
                rows = merge(
                    pages
                        .iter()
                        .filter(|q| !q.decision.skip)
                        .map(|q| q.rows)
                        .collect(),
                );
                steps.push((
                    "page index",
                    if rows.is_empty() {
                        skip("every page's bounds rule it out")
                    } else {
                        read(format!(
                            "{} of {} pages may hold a match",
                            pages.iter().filter(|q| !q.decision.skip).count(),
                            pages.len()
                        ))
                    },
                ));
                skipped |= rows.is_empty();
            }
        }
        let mut reads = Vec::new();
        if !skipped {
            for &column in projection {
                let (r, oi_bytes) = column_read(
                    file,
                    &rg.columns[column],
                    column,
                    &rows,
                    rg.num_rows,
                    use_pages,
                );
                index_bytes += oi_bytes;
                reads.push(r);
            }
        }
        groups.push(RowGroupPlan {
            index: g,
            num_rows: rg.num_rows,
            steps,
            skipped,
            pages,
            rows: if skipped { vec![] } else { rows },
            reads,
            index_bytes,
        });
    }
    Ok(Plan { row_groups: groups })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ranges_merge_when_they_touch() {
        assert_eq!(
            merge(vec![(40, 80), (0, 40), (120, 160)]),
            vec![(0, 80), (120, 160)]
        );
    }
}

File: python/parquet_lab/prune.py
"""Skipping data: deciding, from metadata alone, what a query does not need to read (ch09).

A reader is given a condition on one column, such as ``order_id = 431`` or
``amount_cents > 9000``. Before it reads a page it asks three things, each cheaper than the
reading it may save:

1. **Row group statistics** (:mod:`parquet_lab.stats`), already in the footer: can any value in
   this column chunk's range satisfy the condition?
2. **The Bloom filter** (:mod:`parquet_lab.bloom`), for equality only: was this exact value ever
   added?
3. **The page index** (:mod:`parquet_lab.page_index`): the same question as the first, page by
   page, and then which rows those pages hold, and so which pages of the *other* columns to read.

Every answer is either "skip", which must be certain, or "read", which only means "cannot rule
it out". A wrong "skip" loses rows; a wrong "read" costs bytes. Each decision records why.
"""

from __future__ import annotations

import math
import re
import struct
from dataclasses import dataclass, field
from enum import Enum

from . import bloom, page_index
from .bytes import Span
from .metadata import ColumnChunk, FileMetaData
from .schema import Leaf
from .stats import Comparator, bounds


class Op(Enum):
    EQ = "="
    NOT_EQ = "!="
    LT = "<"
    LT_EQ = "<="
    GT = ">"
    GT_EQ = ">="
    IS_NULL = "is null"
    IS_NOT_NULL = "is not null"

    @staticmethod
    def parse(s: str) -> Op:
        try:
            return Op(s)
        except ValueError:
            raise ValueError(f"unknown comparison {_debug(s)}") from None

    @property
    def symbol(self) -> str:
        return self.value

    def matches(self, ordering: int | None) -> bool:
        """Whether a value compares this way: ``ordering`` is -1, 0 or 1, or ``None`` when the
        value cannot be placed. Null matches nothing but ``is null``."""
        wanted = {
            Op.EQ: (0,),
            Op.NOT_EQ: (-1, 1),
            Op.LT: (-1,),
            Op.LT_EQ: (-1, 0),
            Op.GT: (1,),
            Op.GT_EQ: (1, 0),
        }.get(self, ())
        return ordering in wanted


def _debug(s: str) -> str:
    from .encoding import quote

    return quote(s)


@dataclass
class Predicate:
    """A condition on one column: ``column op value``."""

    column: int
    op: Op
    value: bytes | None
    """The value as PLAIN bytes, as statistics store it. ``None`` for ``is null`` and
    ``is not null``."""
    comparator: Comparator

    @staticmethod
    def new(leaf: Leaf, converted: str | None, op: Op, text: str) -> Predicate:
        """Parse a condition's value for ``leaf``: integers, floats and strings."""
        comparator = Comparator.for_leaf(leaf, converted)
        value = None if op in (Op.IS_NULL, Op.IS_NOT_NULL) else literal(comparator, text.strip())
        return Predicate(leaf.column, op, value, comparator)

    def row_matches(self, value: bytes | None) -> bool:
        """Whether one row's value satisfies the condition. ``None`` is null."""
        if self.op is Op.IS_NULL:
            return value is None
        if self.op is Op.IS_NOT_NULL:
            return value is not None
        if value is None or self.value is None:
            return False
        return self.op.matches(self.comparator.compare(value, self.value))


_INTEGER = re.compile(r"[+-]?[0-9]+")
_FLOAT = re.compile(r"[+-]?(?:[0-9]+\.?[0-9]*|\.[0-9]+)(?:[eE][+-]?[0-9]+)?|[+-]?(?:inf|infinity|nan)", re.I)


def literal(c: Comparator, text: str) -> bytes:
    """A literal as PLAIN bytes for a column compared with ``c``."""
    bad = ValueError(f"{_debug(text)} is not a value of this column")
    ints = {Comparator.I32: ("<i", -(2**31), 2**31), Comparator.U32: ("<I", 0, 2**32),
            Comparator.I64: ("<q", -(2**63), 2**63), Comparator.U64: ("<Q", 0, 2**64)}  # fmt: skip
    if c in ints:
        f, low, high = ints[c]
        if not _INTEGER.fullmatch(text) or not low <= int(text) < high:
            raise bad
        return struct.pack(f, int(text))
    if c in (Comparator.F32, Comparator.F64):
        if not _FLOAT.fullmatch(text):
            raise bad
        v = float(text)
        if c is Comparator.F32:
            # Out of a 32-bit float's range, the value is infinite, as parsing one would make it.
            if math.isfinite(v) and abs(v) > 3.4028235677973366e38:
                v = math.copysign(math.inf, v)
            return struct.pack("<f", v)
        return struct.pack("<d", v)
    if c is Comparator.BYTES:
        return text.strip('"').encode()
    raise ValueError("this reader parses values only for integer, float and string columns")


@dataclass
class Decision:
    """A decision and the reason for it."""

    skip: bool
    why: str


def skip(why: str) -> Decision:
    return Decision(True, why)


def read(why: str) -> Decision:
    return Decision(False, why)


def against_bounds(
    p: Predicate, found: tuple[bytes, bytes] | None, null_count: int | None, num_values: int
) -> Decision:
    """Decide from a minimum and maximum (when usable), a null count, and a value count."""
    all_null = null_count == num_values and num_values > 0
    if p.op is Op.IS_NULL:
        return (
            skip("the null count is zero")
            if null_count == 0
            else read("the null count is not zero, or not recorded")
        )
    if p.op is Op.IS_NOT_NULL:
        return skip("every value is null") if all_null else read("some values are not null")
    if all_null:
        return skip("every value is null, and null satisfies no comparison")
    if p.value is None or found is None:
        return read("no usable minimum and maximum")
    low, high = found
    vs_min, vs_max = p.comparator.compare(p.value, low), p.comparator.compare(p.value, high)
    if vs_min is None or vs_max is None:
        return read("the value cannot be placed in the column's order")
    ruled_out = {
        Op.EQ: vs_min == -1 or vs_max == 1,
        Op.NOT_EQ: vs_min == 0 and vs_max == 0,
        Op.LT: vs_min != 1,  # min >= x
        Op.LT_EQ: vs_min == -1,  # min > x
        Op.GT: vs_max != -1,  # max <= x
        Op.GT_EQ: vs_max == 1,  # max < x
    }[p.op]
    if not ruled_out:
        return read("the range may hold a match")
    if p.op is Op.EQ:
        return skip("the value is below the minimum" if vs_min == -1 else "the value is above the maximum")
    if p.op is Op.NOT_EQ:
        return skip("every value equals it")
    if p.op in (Op.LT, Op.LT_EQ):
        return skip("the minimum is too large")
    return skip("the maximum is too small")


@dataclass(frozen=True)
class Mechanisms:
    """Which mechanisms a plan may use."""

    statistics: bool = True
    bloom: bool = True
    page_index: bool = True


ALL = Mechanisms()
NONE = Mechanisms(False, False, False)


@dataclass
class PagePlan:
    """One page of the condition's column, and whether its rows are kept."""

    span: Span
    rows: tuple[int, int]
    decision: Decision


@dataclass
class ColumnRead:
    """The bytes of one projected column a plan reads in one row group."""

    column: int
    spans: list[Span]
    pages_read: int
    """How many of the chunk's data pages those spans include, and how many it has."""
    pages_total: int


@dataclass
class RowGroupPlan:
    index: int
    num_rows: int
    steps: list[tuple[str, Decision]]
    """Each mechanism consulted, in order, and what it decided."""
    skipped: bool
    pages: list[PagePlan]
    rows: list[tuple[int, int]]
    """The rows still to read, as ``[start, end)`` ranges within the row group."""
    reads: list[ColumnRead]
    index_bytes: int
    """Bytes of Bloom filters and page indexes fetched to decide."""


@dataclass
class Plan:
    row_groups: list[RowGroupPlan] = field(default_factory=list)

    def bytes_read(self) -> int:
        return sum(s.length for g in self.row_groups for r in g.reads for s in r.spans)

    def index_bytes(self) -> int:
        return sum(g.index_bytes for g in self.row_groups)

    def rows_read(self) -> int:
        return sum(b - a for g in self.row_groups for a, b in g.rows)


def _overlaps(a: tuple[int, int], b: tuple[int, int]) -> bool:
    return a[0] < b[1] and b[0] < a[1]


def _merge(ranges: list[tuple[int, int]]) -> list[tuple[int, int]]:
    """Merge touching or overlapping ranges."""
    out: list[tuple[int, int]] = []
    for r in sorted(ranges):
        if out and r[0] <= out[-1][1]:
            out[-1] = (out[-1][0], max(out[-1][1], r[1]))
        else:
            out.append(r)
    return out


def _offset_index_quietly(file: bytes, chunk: ColumnChunk):
    try:
        return page_index.offset_index(file, chunk)
    except ValueError:
        return None


def column_read(
    file: bytes, chunk: ColumnChunk, column: int, rows: list[tuple[int, int]], num_rows: int, use_index: bool
) -> tuple[ColumnRead, int]:
    """The bytes of ``chunk`` needed for ``rows``: its dictionary page, if any, and the data pages
    whose rows overlap them. Without an offset index, the whole chunk."""
    whole = chunk.byte_range()
    oi = _offset_index_quietly(file, chunk) if use_index else None
    if oi is None:
        found = _offset_index_quietly(file, chunk)
        pages = len(found.pages) if found else 0
        return ColumnRead(column, [whole], pages, pages), 0
    spans = []
    # The dictionary page sits before the first data page; any page read needs it.
    first_data = oi.pages[0].offset if oi.pages else whole.end
    ranges = oi.row_ranges(num_rows)
    wanted = [i for i in range(len(oi.pages)) if any(_overlaps(ranges[i], r) for r in rows)]
    if wanted and first_data > whole.start:
        spans.append(Span(whole.start, first_data))
    spans += [oi.pages[i].span() for i in wanted]
    return ColumnRead(column, spans, len(wanted), len(oi.pages)), oi.span.length


def plan(
    file: bytes, md: FileMetaData, leaf: Leaf, p: Predicate, projection: list[int], use: Mechanisms
) -> Plan:
    """Plan a read of ``projection`` for rows satisfying ``p``."""
    orders = md.column_orders
    type_order = orders is not None and leaf.column < len(orders) and orders[leaf.column] == "TYPE_ORDER"
    out = Plan()
    for g, rg in enumerate(md.row_groups):
        chunk = rg.columns[leaf.column]
        steps: list[tuple[str, Decision]] = []
        index_bytes = 0
        skipped = False
        if use.statistics:
            if chunk.statistics is None:
                d = read("the column chunk has no statistics")
            else:
                try:
                    b = bounds(chunk.statistics, p.comparator, type_order)
                    found = (b.min, b.max)
                except ValueError:
                    found = None
                d = against_bounds(p, found, chunk.statistics.null_count, chunk.num_values)
            skipped |= d.skip
            steps.append(("statistics", d))
        if use.bloom and not skipped and p.op is Op.EQ:
            f = bloom.read(file, chunk)
            if f is not None and p.value is not None:
                index_bytes += f.bitset_span.end - f.header_span.start
                d = (
                    read("every bit the value's hash chose is set")
                    if f.probe(p.value).may_contain
                    else skip("a bit the value's hash chose is clear")
                )
                skipped |= d.skip
                steps.append(("Bloom filter", d))
        pages: list[PagePlan] = []
        rows = [] if skipped else [(0, rg.num_rows)]
        # Page rows equal page values only when the column does not repeat; for a repeated column
        # the reader would need the level histograms, and this one does not use them.
        use_pages = use.page_index and not skipped and leaf.max_repetition_level == 0
        if use_pages:
            ci = page_index.column_index(file, chunk)
            oi = page_index.offset_index(file, chunk)
            if ci is not None and oi is not None:
                index_bytes += ci.span.length
                ranges = oi.row_ranges(rg.num_rows)
                for i, loc in enumerate(oi.pages):
                    page_rows = ranges[i][1] - ranges[i][0]
                    found = None
                    if not ci.null_pages[i] and type_order:
                        low, high = ci.min_values[i], ci.max_values[i]
                        if (
                            p.comparator.compare(low, low) is not None
                            and p.comparator.compare(high, high) is not None
                        ):
                            found = (low, high)
                    nulls = (
                        page_rows
                        if ci.null_pages[i]
                        else (ci.null_counts[i] if ci.null_counts is not None else None)
                    )
                    pages.append(PagePlan(loc.span(), ranges[i], against_bounds(p, found, nulls, page_rows)))
                rows = _merge([q.rows for q in pages if not q.decision.skip])
                kept = sum(1 for q in pages if not q.decision.skip)
                steps.append(
                    (
                        "page index",
                        skip("every page's bounds rule it out")
                        if not rows
                        else read(f"{kept} of {len(pages)} pages may hold a match"),
                    )
                )
                skipped |= not rows
        reads = []
        if not skipped:
            for column in projection:
                r, oi_bytes = column_read(file, rg.columns[column], column, rows, rg.num_rows, use_pages)
                index_bytes += oi_bytes
                reads.append(r)
        out.row_groups.append(
            RowGroupPlan(g, rg.num_rows, steps, skipped, pages, [] if skipped else rows, reads, index_bytes)
        )
    return out
